language/move-vm/runtime/src/native_functions.rs

```diff
@@ -139,6 +139,34 @@ impl<'a, 'b> NativeContext<'a, 'b> {
         self.data_store.events()
     }
 
+    pub fn publish_modules(
+        &mut self,
+        modules: Vec<Vec<u8>>,
+        sender: AccountAddress,
+    ) -> PartialVMResult<()> {
+        let compiled_modules = crate::module_validator::validate_and_compile_modules(
+            &modules,
+            sender,
+            self.data_store,
+            self.resolver.loader(),
+        )
+        .map_err(|e| e.to_partial())?;
+
+        // All modules verified, publish them to data cache
+        for (module, blob) in compiled_modules.into_iter().zip(modules.into_iter()) {
+            self.data_store
+                .publish_module(&module.self_id(), blob)
```

Comment:
I think this change should be considered with caution. The VM is very careful to maintain the [invariant](https://github.com/move-language/move/blob/main/language/documentation/spec/vm.md#code-cache) that code cache does not get updated in the middle of execution, and I'm not sure what might break if this invariant doesn't hold. I can't concretely point to things that will go wrong, but in the spirit of being conservative about invariants in safety-critical code I would suggest that we either: 
- use a different scheme like the one suggested by Todd above that preserves the invariant
- hold on this PR until someone further investigates the implications of breaking the invariant + updates the spec accordingly